src/test/ui/or-patterns/or-patterns-syntactic-fail.stderr

```diff
@@ -0,0 +1,112 @@
+error: expected one of `:` or `@`, found `|`
+  --> $DIR/or-patterns-syntactic-fail.rs:28:14
+   |
+LL |     fn fun(A | B: E) {}
+   |              ^ expected one of `:` or `@` here
+
+error: a leading `|` is only allowed in a top-level pattern
+  --> $DIR/or-patterns-syntactic-fail.rs:37:11
+   |
+LL |     let ( | A | B) = E::A;
+   |           ^ help: remove the `|`
+
+error: a leading `|` is only allowed in a top-level pattern
+  --> $DIR/or-patterns-syntactic-fail.rs:38:11
+   |
+LL |     let ( | A | B,) = (E::B,);
+   |           ^ help: remove the `|`
+
+error: a leading `|` is only allowed in a top-level pattern
+  --> $DIR/or-patterns-syntactic-fail.rs:39:11
+   |
+LL |     let [ | A | B ] = [E::A];
+   |           ^ help: remove the `|`
+
+error: a leading `|` is only allowed in a top-level pattern
+  --> $DIR/or-patterns-syntactic-fail.rs:40:13
+   |
+LL |     let TS( | A | B );
+   |             ^ help: remove the `|`
+
+error: a leading `|` is only allowed in a top-level pattern
+  --> $DIR/or-patterns-syntactic-fail.rs:41:17
+   |
+LL |     let NS { f: | A | B };
+   |                 ^ help: remove the `|`
+
+error: a leading `|` is only allowed in a top-level pattern
+  --> $DIR/or-patterns-syntactic-fail.rs:43:11
+   |
+LL |     let ( || A | B) = E::A;
+   |           ^^ help: remove the `||`
+
+error: a leading `|` is only allowed in a top-level pattern
+  --> $DIR/or-patterns-syntactic-fail.rs:44:11
+   |
+LL |     let [ || A | B ] = [E::A];
+   |           ^^ help: remove the `||`
+
+error: a leading `|` is only allowed in a top-level pattern
+  --> $DIR/or-patterns-syntactic-fail.rs:45:13
+   |
+LL |     let TS( || A | B );
+   |             ^^ help: remove the `||`
+
+error: a leading `|` is only allowed in a top-level pattern
+  --> $DIR/or-patterns-syntactic-fail.rs:46:17
+   |
+LL |     let NS { f: || A | B };
+   |                 ^^ help: remove the `||`
```

Comment:
I worry that this looks remarkably close to a valid closure (`|| ->bool { A | B }`) but we can't know for sure if one was intended in a pattern context without more information. Let's leave it as is but I'm making a mental note of it.

Reply:
Alright; Personally I feel that this is clearly what was intended and that it is a stretch to attempt to *destructure* a closure expression in a pattern since that is not an operation that is meaningful to begin with.